File: savo_ws/src/edge/savo_vo/savo_vo/nodes/vo_republisher_node.py
"""ROS 2 node that republishes VO odometry for downstream consumers."""

from __future__ import annotations

import rclpy
from nav_msgs.msg import Odometry
from rclpy.node import Node

from savo_vo.contracts.frame_names import BASE_LINK_FRAME, ODOM_FRAME
from savo_vo.contracts.parameter_names import (
    BASE_FRAME_PARAM,
    ODOM_FRAME_PARAM,
    ODOM_RAW_TOPIC_PARAM,
    ODOM_TOPIC_PARAM,
)
from savo_vo.contracts.topic_names import VO_ODOM_RAW_TOPIC, VO_ODOM_TOPIC
from savo_vo.ros.qos import odometry_qos


class VORepublisherNode(Node):
    def __init__(self) -> None:
        super().__init__("vo_republisher_node")

        self.declare_parameter(ODOM_RAW_TOPIC_PARAM, VO_ODOM_RAW_TOPIC)
        self.declare_parameter(ODOM_TOPIC_PARAM, VO_ODOM_TOPIC)
        self.declare_parameter(ODOM_FRAME_PARAM, ODOM_FRAME)
        self.declare_parameter(BASE_FRAME_PARAM, BASE_LINK_FRAME)

        self._input_topic = self.get_parameter(
            ODOM_RAW_TOPIC_PARAM
        ).get_parameter_value().string_value
        self._output_topic = self.get_parameter(
            ODOM_TOPIC_PARAM
        ).get_parameter_value().string_value
        self._odom_frame = self.get_parameter(
            ODOM_FRAME_PARAM
        ).get_parameter_value().string_value
        self._base_frame = self.get_parameter(
            BASE_FRAME_PARAM
        ).get_parameter_value().string_value

        self._publisher = self.create_publisher(
            Odometry,
            self._output_topic,
            odometry_qos(),
        )
        self.create_subscription(
            Odometry,
            self._input_topic,
            self._on_odom,
            odometry_qos(),
        )

        self.get_logger().info(
            f"VO republisher started: {self._input_topic} -> {self._output_topic}"
        )

    def _on_odom(self, msg: Odometry) -> None:
        msg.header.frame_id = self._odom_frame
        msg.child_frame_id = self._base_frame
        self._publisher.publish(msg)


def main(args: list[str] | None = None) -> None:
    rclpy.init(args=args)
    node = VORepublisherNode()

    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: savo_ws/src/edge/savo_vo/src/vo_republisher_node.cpp
#include "savo_vo/vo_republisher_node.hpp"

#include <functional>

#include "savo_vo/vo_constants.hpp"

namespace savo_vo
{

VORepublisherNode::VORepublisherNode(
  const rclcpp::NodeOptions & options)
: rclcpp::Node("vo_republisher_node", options)
{
  declare_parameters();
  load_parameters();
  create_publishers();
  create_subscribers();

  RCLCPP_INFO(
    get_logger(),
    "VO republisher started: %s -> %s",
    input_topic_.c_str(),
    output_topic_.c_str());
}

rclcpp::QoS VORepublisherNode::odometry_qos()
{
  rclcpp::QoS qos(rclcpp::KeepLast(10));
  qos.reliable();
  qos.durability_volatile();
  return qos;
}

void VORepublisherNode::declare_parameters()
{
  declare_parameter<std::string>(
    constants::kOdomRawTopicParam,
    constants::kVoOdomRawTopic);

  declare_parameter<std::string>(
    constants::kOdomTopicParam,
    constants::kVoOdomTopic);

  declare_parameter<std::string>(
    constants::kOdomFrameParam,
    constants::kOdomFrame);

  declare_parameter<std::string>(
    constants::kBaseFrameParam,
    constants::kBaseFrame);
}

void VORepublisherNode::load_parameters()
{
  input_topic_ = get_parameter(constants::kOdomRawTopicParam).as_string();
  output_topic_ = get_parameter(constants::kOdomTopicParam).as_string();
  odom_frame_ = get_parameter(constants::kOdomFrameParam).as_string();
  base_frame_ = get_parameter(constants::kBaseFrameParam).as_string();

  if (input_topic_.empty()) {
    input_topic_ = constants::kVoOdomRawTopic;
  }

  if (output_topic_.empty()) {
    output_topic_ = constants::kVoOdomTopic;
  }

  if (odom_frame_.empty()) {
    odom_frame_ = constants::kOdomFrame;
  }

  if (base_frame_.empty()) {
    base_frame_ = constants::kBaseFrame;
  }
}

void VORepublisherNode::create_publishers()
{
  odom_pub_ = create_publisher<Odometry>(
    output_topic_,
    odometry_qos());
}

void VORepublisherNode::create_subscribers()
{
  odom_sub_ = create_subscription<Odometry>(
    input_topic_,
    odometry_qos(),
    std::bind(&VORepublisherNode::on_odom, this, std::placeholders::_1));
}

void VORepublisherNode::on_odom(const Odometry::SharedPtr msg)
{
  if (!msg) {
    return;
  }

  Odometry output = *msg;
  output.header.frame_id = odom_frame_;
  output.child_frame_id = base_frame_;

  odom_pub_->publish(output);
}

}  // namespace savo_vo
